include/rabbitmq-c/amqp.h

```diff
@@ -2433,6 +2433,26 @@ AMQP_EXPORT
 int AMQP_CALL amqp_set_rpc_timeout(amqp_connection_state_t state,
                                    const struct timeval *timeout);
 
+/**
+ * amqp_publisher_confirm_wait
+ *
+ * Wait for a publisher confirm when the connection is in select mode.
+ * If the response has a `reply_type` of `AMQP_RESPONSE_LIBRARY_EXCEPTION` _and_
+ * the `library_error` is `AMQP_STATUS_UNEXPECTED_STATE`, then the frame
+ * received was likely not an ack.
+ *
+ * \param [in] state connection state
+ * \param [in] timeout when waiting for the frame. Passing NULL will result in
+ * blocking behavior
+ * \param [out] pointer to where the envelope details should go
+ * \param [out] pointer to where the ACK details should go
+ * \returns amqp_rpc_reply_t *
+ */
+AMQP_EXPORT
+amqp_rpc_reply_t AMQP_CALL amqp_publisher_confirm_wait(
```

Comment:
Thinking through this API a bit more we need to consider a couple different scenarios:

1. What does this API do if the connection isn't in publisher-confirm mode? From an API-use perspective, it'd be helpful to return some kind of error if the connection isn't correctly configured.
2. In publisher confirm mode, when a message is published the broker will return one of basic.ack, basic.nack, basic.reject (with channel.close and connection.close as options, but already handled). (see: https://github.com/alanxz/SimpleAmqpClient/blob/500b070465f7759f0bf984ff56b724c06df930ef/src/Channel.cpp#L837-L877). From API-use perspective, it'd be helpful if this function handled all 3 cases, likely this means: this function should have as an out parameter an `amqp_method_number_t`, and a `void*`, which is a library allocated version of one of the previously referenced methods.

Reply:
Since this new function doesn't own the memory of the decoded payload, it's creating non-fun memory safety issues. I'm thinking of just making a simple union of the different method response types, and having that be an out param instead of the `void*`, and then consolidating all of the out params into one out param.

Something like:

```c
typedef union amqp_publisher_confirm_payload_t_ {
  amqp_basic_ack_t ack;
  amqp_basic_nack_t nack;
  amqp_basic_reject_t reject;
} amqp_publisher_confirm_payload_t;

typedef struct amqp_publisher_confirm_t_ {
  amqp_publisher_confirm_payload_t payload;
  amqp_channel_t channel;
  amqp_method_t method;
} amqp_publisher_confirm_t;
```

Then I can change the function signature to:

```c
amqp_rpc_reply_t AMQP_CALL amqp_publisher_confirm_wait(
    amqp_connection_state_t state, const struct timeval *timeout,
    amqp_publisher_confirm_t *response);
```

That kinda seems better to me, what are your thoughts? This can allow the caller to own all of the memory, and only marginally increases the complexity of this function.